proxy/src/main/java/com/velocitypowered/proxy/plugin/loader/JavaPluginLoader.java

```diff
@@ -0,0 +1,123 @@
+import com.velocitypowered.proxy.plugin.loader.java.JavaVelocityPluginDescription;
+
+import javax.annotation.Nonnull;
+import java.io.BufferedInputStream;
+import java.lang.annotation.Annotation;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.jar.Attributes;
+import java.util.jar.JarInputStream;
+import java.util.jar.Manifest;
+
+public class JavaPluginLoader implements PluginLoader {
+    private final ProxyServer server;
+
+    public JavaPluginLoader(ProxyServer server) {
+        this.server = server;
+    }
+
+    @Nonnull
+    @Override
+    public PluginDescription loadPlugin(Path source) throws Exception {
+        String mainClassName = getMainClassName(source);
+
+        if (mainClassName == null) {
+            throw new AssertionError();
+        }
+
+        PluginClassLoader loader = new PluginClassLoader(
+                new URL[] {source.toUri().toURL() }
+        );
+
+        Class mainClass = loader.loadClass(mainClassName);
+        Annotation annotation = mainClass.getAnnotation(Plugin.class);
+
+        if (!(annotation instanceof Plugin)) {
+            throw new InvalidPluginException("Main class does not have @Plugin annotation");
+        }
+
+        return createDescription((Plugin) annotation, source, mainClass);
+    }
+
+    @Nonnull
+    @Override
+    public PluginContainer createPlugin(PluginDescription description) throws Exception {
+        if (!(description instanceof JavaVelocityPluginDescription)) {
+            throw new IllegalArgumentException("Description provided isn't of the Java plugin loader");
+        }
+
+        JavaVelocityPluginDescription javaDescription = (JavaVelocityPluginDescription) description;
+        Optional<Path> source = javaDescription.getSource();
+
+        if (!source.isPresent()) {
+            throw new IllegalArgumentException("No path in plugin description");
+        }
+
+        Object instance = javaDescription.getMainClass().newInstance();
+        // TODO Inject server variable, logger...
+        // Should we add Guice?
+
+        return new VelocityPluginContainer(
+            description.getId(),
+            description.getVersion().orElse(null),
+            description.getAuthor().orElse(null),
+            description.getDependencies(),
+            source.get(),
+            instance
+        );
+    }
+
+    private String getMainClassName(Path source) throws Exception {
+        try (JarInputStream in = new JarInputStream(new BufferedInputStream(Files.newInputStream(source)))) {
+            Manifest manifest = in.getManifest();
+
+            if (manifest == null) {
+                throw new InvalidPluginException("Jar does not contain a manifest");
+            }
+
+            Attributes attributes = manifest.getMainAttributes();
+            String className = attributes.getValue("Main-Class");
```

Comment:
This isn't a good idea. We shouldn't be using any existing Java entries.

Reply:
Yeah, we've been discussing on Discord about removing the option to have multiple plugins on one Jar file.

I agree that the `Main-Class` attribute was an ugly idea, will change it to a `.toml` file.